### PR TITLE
Do you handle multi os dev teams

### PR DESCRIPTION
**Pull request by tomek-i**
>
> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️ BizCover client work
https://github.com/SSWConsulting/SSW.SAW.General/issues/4

> 2. What was changed?

✏️ created new rule

> 3. Did you do pair or mob programming (list names)?

✏️🥷🏻

### CODE REVIEW

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
+---
+seoDescription: Learn how to manage a multi-OS development team in source control. This article covers best practices for handling line endings (CRLF vs LF), creating cross-platform setup scripts, configuring Git for consistency, and using .NET CLI for secrets management to boost efficiency across Windows, macOS, and Linux.
+type: rule
+title: Do you handle Multi-OS dev teams in source control?
+uri: do-you-handle-multi-os-dev-teams-in-source-control
```
Comment (tiagov8) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:5]: Hi @tomek-i 

As per other rules,

- [ ] 1. Revome the "do-you-" from URLs (includes folder name)

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+authors:
+  - title: Tom Iwainski
+    url: https://www.ssw.com.au/people/thomas-iwainski
+related: []
+created: 2024-11-05T12:52:08.000Z
+archivedreason: null
+guid: 7bc2e9c8-2b12-466b-8ecd-34576e3a1648
+---
+
+In today’s development world, teams often consist of developers working on different operating systems (OS), such as Windows, macOS, and Linux. While each OS has its own strengths, managing a cross-platform development environment can introduce challenges. Issues like inconsistent line endings, platform-specific setup scripts, and configuration mismatches can lead to headaches for both individual developers and the team as a whole.
```
Comment (tiagov8) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:15]:
```suggestion
In today’s development world, teams often consist of developers working on different operating systems (OS), such as Windows, macOS, and Linux. While each OS has its own strengths, managing a cross-platform development environment can introduce challenges. 

Issues like inconsistent line endings, platform-specific setup scripts, and configuration mismatches can lead to headaches for both individual developers and the team as a whole.
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+## Addressing Line Ending differences (CRLF vs. LF)
+
+One of the most common issues faced by teams working with Git across different platforms is the handling of line endings. Windows uses **CRLF** (Carriage Return + Line Feed) for line endings, while macOS and Linux use **LF** (Line Feed) only. This can lead to unnecessary diffs in Git and potential merge conflicts.
+
+### Solution: Use Git’s Line Ending configuration
+
+Git provides a way to manage line endings across different operating systems by using the `core.autocrlf` setting. This configuration ensures that line endings are normalized when files are checked in and out of the repository.
+
+- **Windows Users**: Set Git to automatically convert line endings to **CRLF** when checking out files, and convert them back to **LF** when committing.
```
Comment (tiagov8) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:27]:
```suggestion
- **Windows users**: Set Git to automatically convert line endings to **CRLF** when checking out files, and convert them back to **LF** when committing.
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  ```bash
+  git config --global core.autocrlf true
+  ```
+
+- **macOS/Linux Users**: Set Git to automatically convert **CRLF** line endings to **LF** when checking out files and leave them as **LF** when committing.
```
Comment (tiagov8) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:33]:
```suggestion
- **macOS/Linux users**: Set Git to automatically convert **CRLF** line endings to **LF** when checking out files and leave them as **LF** when committing.
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+  ```bash
+  git config --global core.autocrlf input
+  ```
+
+- **Repository-Wide Configuration**: It’s a good practice to enforce this configuration across the team via a `.gitattributes` file, which allows you to define how specific file types should be handled. For example:
```
Comment (tiagov8) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:39]:
```suggestion
- **Repository-Wide configuration**: It's a good practice to enforce this configuration across the team via a `.gitattributes` file, which allows you to define how specific file types should be handled. For example:
```

**Review by tiagov8 — CHANGES_REQUESTED** [review]

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  ```text
+  * text=auto
+  *.cs text eol=lf
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:43]: I don't think this is right for `*.cs` files. We shouldn't have any specific configuration. I agree with the settings for `*.sh` and `*.bat`.

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  *.bat text eol=crlf
+  ```
+
+This ensures that, no matter what OS a developer is using, files are checked out and committed with consistent line endings. The `eol` attribute specifically handles cases like batch scripts or shell scripts that may need different line endings.
+
+## Creating Cross-Platform Setup Scripts for Easy Onboarding
+
+Onboarding new developers is a critical step in ensuring that everyone is up and running quickly, but multi-OS teams often struggle with platform-specific setup instructions. The goal is to make onboarding as seamless as possible, whether a developer is using Windows, macOS, or Linux.
+
+### Solution: Write Cross-Platform setup scripts
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:54]: I think this section needs to be re-written. Powershell is cross platform, and should work across all OS's. Therefore the suggestion should just be 'use powershell' 😄

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  ```text
+  .DS_Store
+  Thumbs.db
+  ```
+
+- **Hooks and Templates**: Some repositories might require hooks or commit templates to enforce conventions like conventional commit messages or certain commit checks. Using a `.githooks` or `.gitmessage` file in the repository can help maintain consistency across platforms.
+
+---
+
+## Handling .NET secrets across platforms
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:143]: I don't think secrets has anything to do with git configuration. Consider removing this section.

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  }
+```
+
+This script checks the platform and runs the appropriate setup script based on the OS.
+
+## Consistent Git configuration for Multi-OS Teams
+
+To further ensure smooth collaboration among multi-OS teams, it’s important to standardize Git configurations. Beyond line endings, Git offers other configurations that help maintain consistency.
+
+### Key Git cnfigurations for Multi-OS Teams:
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:117]:
```suggestion
### Key Git configurations for Multi-OS Teams:
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  dotnet user-secrets get "MySecret"
+  ```
+
+By relying on the `.NET CLI`, your team ensures that secrets are handled consistently, regardless of the operating system.
+
+For more information visit [Safe storage of app secrets in development in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=linux)
+
+---
+
+## General best practices for Cross-Platform efficiency
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:169]: IDE's aren't very related to source control. Consider dropping this section.

**Review by danielmackay — CHANGES_REQUESTED** [review]
Few comments

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+type: rule
+title: Do you handle Multi-OS dev teams in source control?
+uri: do-you-handle-multi-os-dev-teams-in-source-control
+authors:
+  - title: Tom Iwainski
+    url: https://www.ssw.com.au/people/thomas-iwainski
+related: []
+created: 2024-11-05T12:52:08.000Z
+archivedreason: null
+guid: 7bc2e9c8-2b12-466b-8ecd-34576e3a1648
```
Comment (danielmackay) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:12]: Rule needs to be linked to a category.

**Review by danielmackay — CHANGES_REQUESTED** [review]
Added another comment

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 6 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:43]: even for *.cs files I am getting Visual Studio always popping up with the file ending has changed `do you want overwrite it?` and that is usually when our MacOS people touched a file :-D

I can take it out though, let me know

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 7 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:54]: oOh i thought pwsh was a 3rd party thing that executes powershell scripts :-D

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 10 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md]: ✅ done

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 8 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:133]: It has nothing to do with git configuration. It has a lot to do with multi os dev teams as mac users will use VS code for development and they have to manually set the secrets via command line while in VStudio you can use the context menu on the project in the IDE.  (not sure about rider)

As most people will be using VStudio and then write the documentation targeting this IDE they need to figure out the command - it is much better practice in the onboarding documentation to just write the specific command that the devs can copy and paste.

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+- **Hooks and Templates**: Some repositories might require hooks or commit templates to enforce conventions like conventional commit messages or certain commit checks. Using a `.githooks` or `.gitmessage` file in the repository can help maintain consistency across platforms.
+
+---
+
+## Handling .NET secrets across platforms
+
+When working with .NET applications, developers often use the Visual Studio UI for managing secrets. The UI doesn’t work well across different platforms. To ensure consistency, the **CLI** is a better choice for managing secrets in a cross-platform environment.
+
+### Solution: Use the .NET CLI for secrets management
```
Comment (tomek-i) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:137]:
```suggestion
### Solution: Use the .NET CLI for local development secrets management
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+  Thumbs.db
+  ```
+
+- **Hooks and Templates**: Some repositories might require hooks or commit templates to enforce conventions like conventional commit messages or certain commit checks. Using a `.githooks` or `.gitmessage` file in the repository can help maintain consistency across platforms.
+
+---
+
+## Handling .NET secrets across platforms
+
+When working with .NET applications, developers often use the Visual Studio UI for managing secrets. The UI doesn’t work well across different platforms. To ensure consistency, the **CLI** is a better choice for managing secrets in a cross-platform environment.
```
Comment (tomek-i) [inline rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:135]:
```suggestion
When working with .NET applications, developers often use the Visual Studio UI for managing local development secrets. The UI doesn’t work well across different platforms or IDE's like VS Code. To ensure consistency, the **CLI** is a better choice for managing secrets in a cross-platform environment.
```

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 7 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md]: ✅ done

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 1 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md]: ✅ done

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 11 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:12]: Which one?

Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:12]: ✅ done. rules-to-better-code

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 6 above)
Comment (danielmackay) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:45]: I still don't think it's needed. I believe the problem is elsewhere.

Have a read of this post from Scott H: https://www.hanselman.com/blog/youre-just-another-carriage-return-line-feed-in-the-wall

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 8 above)
Comment (danielmackay) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:133]: Most MacOS users will use Rider for .NET Development. I don't know anyone who solely uses VS Code for .NET.

What you mention above is correct, but the rule is about source control and I don't think this comment belongs here..

File: categories/software-engineering/rules-to-better-code.md [hunk 14]
```diff
@@ -90,6 +90,7 @@ index:
   - handle-unhappy-paths
   - get-code-line-metrics
   - single-codebase-for-multiple-domains-with-tinacm-nextjs
+  - do-you-handle-multi-os-dev-teams-in-source-control
```
Comment (danielmackay) [inline categories/software-engineering/rules-to-better-code.md:93]: Can you remove the rule from this category and instead add to these two:

- https://www.ssw.com.au/rules/rules-to-better-version-control-with-tfs-aka-source-control/
- https://www.ssw.com.au/rules/rules-to-better-version-control-with-git/

**Review by danielmackay — CHANGES_REQUESTED** [review]
1 new comment and 2 follow ups on existing conversations.

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 6 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:45]: ✅ Done - removed it.

File: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md (hunk 8 above)
Comment (tomek-i) [reply rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md:133]: ✅ Done.

People who do not have Rider but a MacOS :-D there are some at SSW

File: categories/software-engineering/rules-to-better-code.md (hunk 14 above)
Comment (tomek-i) [reply categories/software-engineering/rules-to-better-code.md:93]: ✅ Done

**Comment by tomek-i** [comment]
@danielmackay applied all changes

**Review by danielmackay — APPROVED** [review]
LGTM

**Pull request merged**